Let `edm::Wrapper<T>` use the constructor `T{kUninitialized}`

fwyzard opened the pull request:
#### PR description:

Let `edm::Wrapper<T>` use a special constructor `T{kUninitialized}` instead of the default constructor `T{}`, if the latter does not exist.

Add to `DeviceProduct<T>` an explicit constructor `DeviceProduct{kUninitialized}` that calls `T{kUninitialized}`, if the latter is valid.

Remove the default constructor from `PortableCollection`, `PortableObject` and related types, and introduce the special constructor taking `kUninitialized`.


#### PR validation:

Unit tests pass, and the full 2024 HLT runs.

fwyzard committed:
Let edm::Wrapper<T> use the constructor T{kUninitialized}

Let edm::Wrapper<T> use a special constructor T{kUninitialized} instead of
the default constructor T{}, if the latter does not exist.

Extend the edm::Wrapper<T> unit tests to cover this case.

fwyzard committed:
Remove the default constructor from PortableCollection

Remove the default constructor from PortableCollection, PortableObject and
related types, and introduce the special constructor taking kUninitialized.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"2838418ecb9baaf825f0785473915ab844132497":{"files":["DataFormats/Common/interface/AllocateForOverwrite.h","DataFormats/Common/interface/Wrapper.h"],"squashed":false,"time":1733385366},"f7294499f9d0bc00f3369610b1f3a397914c0bff":{"files":["DataFormats/Common/interface/DeviceProduct.h","DataFormats/Portable/interface/PortableDeviceCollection.h","DataFormats/Portable/interface/PortableDeviceObject.h","DataFormats/Portable/interface/PortableHostCollection.h","DataFormats/Portable/interface/PortableHostObject.h","DataFormats/SiPixelClusterSoA/interface/SiPixelClustersDevice.h","DataFormats/SiPixelClusterSoA/interface/SiPixelClustersHost.h","DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsDevice.h","DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsHost.h","DataFormats/SiPixelDigiSoA/interface/SiPixelDigisDevice.h","DataFormats/SiPixelDigiSoA/interface/SiPixelDigisHost.h","DataFormats/TrackSoA/interface/TracksDevice.h","DataFormats/TrackSoA/interface/TracksHost.h","DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsDevice.h","DataFormats/TrackingRecHitSoA/interface/TrackingRecHitsHost.h"],"squashed":false,"time":1733394700}},"emoji":{},"last_seen_sha":"f7294499f9d0bc00f3369610b1f3a397914c0bff","signatures":{}} -->

fwyzard commented:
please test with #46876, #46875

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42905

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/interface/DeviceProduct.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsDevice.h  modified in PR(s): #44844
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigisDevice.h  modified in PR(s): #44844

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42905/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42905/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42907

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/interface/DeviceProduct.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsDevice.h  modified in PR(s): #44844
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigisDevice.h  modified in PR(s): #44844

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42907/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42907/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
please test with #46876, #46875

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42908

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Common/interface/DeviceProduct.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableDeviceCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostCollection.h  modified in PR(s): #44844
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigiErrorsDevice.h  modified in PR(s): #44844
  - File  DataFormats/SiPixelDigiSoA/interface/SiPixelDigisDevice.h  modified in PR(s): #44844

cmsbuild commented:
A new Pull Request was created by  @fwyzard for master.

It involves the following packages:

- DataFormats/Common (**core**)
- DataFormats/Portable (**heterogeneous**)
- DataFormats/SiPixelClusterSoA (**reconstruction, heterogeneous**)
- DataFormats/SiPixelDigiSoA (**reconstruction, heterogeneous**)
- DataFormats/TrackSoA (**reconstruction, heterogeneous**)
- DataFormats/TrackingRecHitSoA (**reconstruction, heterogeneous**)


@Dr15Jones, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please review it and eventually sign? Thanks.
@VinInn, @VourMa, @dkotlins, @ferencek, @gpetruc, @makortel, @missirol, @mmusich, @mroguljic, @mtosi, @rovere, @tsusa, @tvami, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented:
please test with #46876, #46875

fwyzard commented:
please test with #46876, #46875

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2be352/43285/summary.html
**COMMIT**: 96b3c3c2386fad8cd17668ec62f862d4d60245f0
**CMSSW**: CMSSW_15_0_X_2024-12-05-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46877/43285/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 3 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-05-1100+2be352/66000/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3606632
 - DQMHistoTests: [Total failures: 518](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-05-1100+2be352/66000/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3606094
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 206 log files, 177 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-05-1100+2be352/66000/triggerResults): found differences in 1 / 45 workflows

makortel reviewed (commented):
The overall pattern looks good, thanks.

The new behavior of `edm::Wrapper<T>` would need tests in the framework code:
* In [DataFormats/Common/test/Wrapper_t.cpp](https://github.com/cms-sw/cmssw/blob/master/DataFormats/Common/test/Wrapper_t.cpp) minimally that `T` without a default constructor compiles
* I think a test that a `T` without default constructor also works through (ROOT) IO, e.g. along
  * Add an int-holding type in that does not have a default constructor to [DataFormats/TestObjects/interface/ToyProducts.h](https://github.com/cms-sw/cmssw/blob/master/DataFormats/TestObjects/interface/ToyProducts.h)
  * Add a producer and a testing consumer to [FWCore/Integration/plugins](https://github.com/cms-sw/cmssw/tree/master/FWCore/Integration/plugins), and test configurations to [FWCore/Integration/test](https://github.com/cms-sw/cmssw/tree/master/FWCore/Integration/test)

If you feel like it, adding minimal unit tests for `edm::DeviceProduct<T>` for `T`'s with and without a default constructor (similarly to the `Wrapper_t.cpp`) would be awesome, but since `DeviceProduct<T>` doesn't have any tests yet (my bad), I'm not requiring those in this PR (and I can add those later instead).

fwyzard commented on DataFormats/Common/interface/DeviceProduct.h:
```diff
@@ -45,7 +47,7 @@ namespace edm {
   class DeviceProduct : public DeviceProductBase {
   public:
-    DeviceProduct() = default;
+    explicit DeviceProduct(edm::AllocateForOverwrite) : data_{edm::allocateForOverwrite} {}
```
> (just to note, if we'd really want to, we could allow T to not have neither default or AllocateForOverwrite constructor)

how would that work ?
don't we need `edm::Wrapper<T>` to be constructible either by default or passing `kUninitialized` ?

fwyzard replied:
OK, I've changed `DeviceProduct` to support products that have either a default constructor, o a constructor that take a `kUninitialized`.

I've tested it reverting locally some of the changes to the device dataformats to use the default constructor, and it seems to work.

I thought about adding some unit tests, but I realised that I have no idea what we should test for :-/

makortel replied:
> > (just to note, if we'd really want to, we could allow T to not have neither default or AllocateForOverwrite constructor)
> 
> how would that work ? don't we need `edm::Wrapper<T>` to be constructible either by default or passing `kUninitialized` ?

_Theoretically_ we could be able use `std::optional<T>` for `data_` member instead of `T` when `T` does not have either default or `edm::Uninitialized` constructor. But maybe it's not worth of the hassle.

fwyzard replied:
Ah, I see... but if we do that couldn't we drop the `edm::Uninitialized` part completely ?

makortel replied:
> Ah, I see... but if we do that couldn't we drop the `edm::Uninitialized` part completely ?

You're right, I'd think in that case the `edm::Uninitialized` constructor could be omitted. I'd consider also making the default constructor of `DeviceProduct` (or probably the `DeviceProductBase`) to throw an exception (or `assert()`) because it should never be called in practice.

fwyzard commented on DataFormats/Common/interface/AllocateForOverwrite.h:
```diff
@@ -0,0 +1,19 @@
+
+  struct AllocateForOverwrite {};
+
+  static const constexpr inline AllocateForOverwrite allocateForOverwrite;
```
ok

fwyzard commented on DataFormats/Common/interface/AllocateForOverwrite.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace edm {
+
+  struct AllocateForOverwrite {};
```
ok

fwyzard commented on DataFormats/Common/interface/Wrapper.h:
```diff
@@ -49,6 +50,14 @@ namespace edm {
+      } else {
+        return T(edm::allocateForOverwrite);
+      }
+    }
```
> However, reversing the conditional ... would ... lead to compilation error

Yes, that's what I tried first, and I run into this kind of errors.

I'm happy to try and implement this inverted approach if you think it worthwhile.

makortel replied:
Maybe we leave it to a later time (usefulness in practice is somewhat unclear, and a straightforward approach wouldn't work well with third party classes).

fwyzard committed:
Add tests for edm::Wrapper<T> using T{kUninitialized}

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42935

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42935/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42935/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented:
I've renamed the new type and added the `edm::Wrapper` tests, I'll add those for `DeviceProduct` in the next commits.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42936

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

cmsbuild commented:
Pull request #46877 was updated. @Dr15Jones, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 64KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2be352/43306/summary.html
**COMMIT**: 6e1a47d331aa4ddba83d2837b41d1980b1b9ae49
**CMSSW**: CMSSW_15_0_X_2024-12-06-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46877/43306/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIOPoolInputNoParentDictionary had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-06-2300+2be352/66013/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 56 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-06-2300+2be352/66013/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3623082
 - DQMHistoTests: [Total failures: 2289](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-06-2300+2be352/66013/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3620773
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 206 log files, 177 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-06-2300+2be352/66013/triggerResults): no differences found

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42939

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

cmsbuild commented:
Pull request #46877 was updated. @Dr15Jones, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

fwyzard committed:
Let DeviceProduct<T> use the constructor T{kUninitialized}

Let DeviceProduct<T> use a special constructor T{kUninitialized} instead of the
default constructor T{}, if it is available.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42941

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

cmsbuild commented:
Pull request #46877 was updated. @Dr15Jones, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

fwyzard commented:
+heterogeneous

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2be352/43310/summary.html
**COMMIT**: 59dc5fc017d38552662f1552c20f04ca6206d2d5
**CMSSW**: CMSSW_15_0_X_2024-12-07-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46877/43310/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIOPoolInputNoParentDictionary had ERRORS
</pre>

## Comparison Summary



There are some workflows for which there are errors in the baseline:
 1000.0 step 4
10224.0 step 3
11634.0 step 3
12434.0 step 3
12434.7 step 3
12834.0 step 3
12834.7 step 3
12846.0 step 3
13034.0 step 3
1306.0 step 3
13234.0 step 2
1330.0 step 3
135.4 step 1
136.731 step 4
136.793 step 4
136.874 step 4
139.001 step 4
140.56 step 3
14034.0 step 2
14234.0 step 2
16834.0 step 3
2022.000001 step 4
2022.002001 step 4
2023.002001 step 4
2024.000001 step 4
2024.101001 step 4
2024.202001 step 4
2024.303001 step 4
24834.911 step 3
25.0 step 3
2500.201 step 3
250202.181 step 4
25202.0 step 3
29634.0 step 3
29634.911 step 3
29696.0 step 3
29700.0 step 3
29834.999 step 4
312.0 step 3
4.22 step 5
4.53 step 4
5.1 step 1
7.3 step 5
8.0 step 5
9.0 step 3
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

**Summary**:
 - You potentially [added 3095 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-07-1100+2be352/66015/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-07-1100+2be352/66015/validateJR.html)
 - DQMHistoTests: Total files compared: 2
 - DQMHistoTests: Total histograms compared: 456
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-07-1100+2be352/66015/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 456
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 1 files compared)
 - Checked 162 log files, 109 edm output root files, 2 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-07-1100+2be352/66015/triggerResults): no differences found

fwyzard committed:
Fix IOPool/Input test for modified DataFormats/Common

Fix the TestIOPoolInputNoParentDictionary test to build correctly even
if DataFormats/Common was modified in the local release area.

fwyzard committed:
Rename IOPool/Input test for consistency

Rename run_TestRefProductIDMetadataConsistencyRoot.sh to
testRefProductIDMetadataConsistencyRoot.sh to match the other tests.

fwyzard commented:
please test

fwyzard commented:
+heterogeneous

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42942

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

cmsbuild commented:
Pull request #46877 was updated. @Dr15Jones, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

fwyzard commented:
> ---> test TestIOPoolInputNoParentDictionary had ERRORS

The failure is mostly unrelated to these changes, other than the fact that any change to `DataFormats/Common` that is not compatible with the base release would make the test fail.

I've tried to make the test more robust, and not it should pass as well.

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2be352/43319/summary.html
**COMMIT**: 1532a8d5e88477a1613d8a063962aaa08b0f9f05
**CMSSW**: CMSSW_15_0_X_2024-12-08-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46877/43319/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-08-0000+2be352/66024/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 318 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-08-0000+2be352/66024/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3623082
 - DQMHistoTests: [Total failures: 3722](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-08-0000+2be352/66024/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3619340
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 206 log files, 177 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-08-0000+2be352/66024/triggerResults): no differences found

jfernan2 commented:
+1

makortel commented on DataFormats/TestObjects/src/classes_def.xml:
```diff
@@ -5,7 +5,9 @@
  <class name="edmtest::IntProduct" ClassVersion="10">
   <version ClassVersion="10" checksum="4286676158"/>
  </class>
- <class name="std::pair<edmtest::IntProduct, edm::RandomNumberGeneratorState>"/>
```
Why was this dictionary removed?

fwyzard replied:
Actually, no idea, so I guess "by mistake".

fwyzard commented:
@iarspider @smuzaffar my last commit modified only `DataFormats/TestObjects/src/classes_def.xml`, which is only in the @cms-sw/core-l2 category.

Why did the bot reset also the other signatures ?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46877/42962

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/Portable/interface/PortableHostObject.h  modified in PR(s): #43969

fwyzard commented:
+heterogeneous

fwyzard commented:
please test

cmsbuild commented:
Pull request #46877 was updated. @Dr15Jones, @jfernan2, @makortel, @mandrenguyen, @smuzaffar can you please check and sign again.

smuzaffar commented:
> @iarspider @smuzaffar my last commit modified only `DataFormats/TestObjects/src/classes_def.xml`, which is only in the @cms-sw/core-l2 category.
> 
> Why did the bot reset also the other signatures ?

you pushed 3 commits few minutes ago and https://github.com/cms-sw/cmssw/pull/46877/commits/4c8434cf1594cdc109b4b5c5cd78eefece651d29 is touch non-core files

fwyzard commented:
> you pushed 3 commits few minutes ago and [4c8434c](https://github.com/cms-sw/cmssw/commit/4c8434cf1594cdc109b4b5c5cd78eefece651d29) is touching non-core files

The commits touch those packages, but [the comparison](https://github.com/cms-sw/cmssw/compare/1532a8d5e88477a1613d8a063962aaa08b0f9f05..4c8434cf1594cdc109b4b5c5cd78eefece651d29) does not:
#### `DataFormats/TestObjects/src/classes_def.xml`
```diff
@@ -5,6 +5,7 @@
 <class name="edmtest::IntProduct" ClassVersion="10">
  <version ClassVersion="10" checksum="4286676158"/>
 </class>
+<class name="std::pair<edmtest::IntProduct, edm::RandomNumberGeneratorState>"/>
 <class name="edmtest::MaybeUninitializedIntProduct" ClassVersion="3">
  <version ClassVersion="3" checksum="3324182225"/>
 </class>
```

I though that the bot would consider the actual comparison, not the commits, because we want people to be able to replace the commits with a squashed version without requiring to re-sign the PRs ?

smuzaffar commented:
> I though that the bot would consider the actual comparison

No bot uses the commits and signatures are assigned to commits.

fwyzard commented:
Ah, I see.

Could we consider to use the comparison instead of the commits as a future improvement ?

smuzaffar commented:
I do not think it is possible specially when one force push changes and ref to old commits are gone

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2be352/43357/summary.html
**COMMIT**: 4c8434cf1594cdc109b4b5c5cd78eefece651d29
**CMSSW**: CMSSW_15_0_X_2024-12-10-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46877/43357/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+2be352/66059/validateJR.html)
 - DQMHistoTests: Total files compared: 47
 - DQMHistoTests: Total histograms compared: 3623082
 - DQMHistoTests: [Total failures: 473](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+2be352/66059/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3622589
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 46 files compared)
 - Checked 206 log files, 177 edm output root files, 47 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+2be352/66059/triggerResults): found differences in 1 / 45 workflows

makortel commented:
> I do not think it is possible specially when one force push changes and ref to old commits are gone

Darn. Although https://github.com/cms-sw/cmssw/issues/43845 lists 
> * Not reset L2's signatures as long as the changes proposed by a force update are identical to changes before the force push. This includes
> 
>   * squashing multiple commits (`c1, c2, c3, ..., cn`) in to one (or more) new commit(s) (`cx`)
>   * force pushing a different set of commits which have overall identical changes

So did that not work in the end?

makortel commented:
+core

smuzaffar commented:
> So did that not work in the end?

@iarspider can check. I think squashing commits is working. I am not sure about adding new set of commits with identical changes though

iarspider commented:
The changes were not identical: [before](https://github.com/cms-sw/cmssw/compare/87f3f24501aefe75d4e0be6f3f624e890b8a621d...1532a8d5e88477a1613d8a063962aaa08b0f9f05#diff-5ada35958f5fceb1401a6c598a97452b0cee26cca7cd4612e2fa0dea23d7b575) and [after](https://github.com/cms-sw/cmssw/compare/87f3f24501aefe75d4e0be6f3f624e890b8a621d...4c8434cf1594cdc109b4b5c5cd78eefece651d29#diff-5ada35958f5fceb1401a6c598a97452b0cee26cca7cd4612e2fa0dea23d7b575) squash.

smuzaffar commented:
lets move this discussion to https://github.com/cms-sw/cmssw/issues/43845

fwyzard commented:
> The changes were not identical: [before](https://github.com/cms-sw/cmssw/compare/87f3f24501aefe75d4e0be6f3f624e890b8a621d...1532a8d5e88477a1613d8a063962aaa08b0f9f05#diff-5ada35958f5fceb1401a6c598a97452b0cee26cca7cd4612e2fa0dea23d7b575) and [after](https://github.com/cms-sw/cmssw/compare/87f3f24501aefe75d4e0be6f3f624e890b8a621d...4c8434cf1594cdc109b4b5c5cd78eefece651d29#diff-5ada35958f5fceb1401a6c598a97452b0cee26cca7cd4612e2fa0dea23d7b575) squash.

As I wrote [above](https://github.com/cms-sw/cmssw/pull/46877#issuecomment-2532283092), [the changes](https://github.com/cms-sw/cmssw/compare/1532a8d5e88477a1613d8a063962aaa08b0f9f05..4c8434cf1594cdc109b4b5c5cd78eefece651d29) between the two set of commits touch only the `DataFormats/TestObjects` package:
#### `DataFormats/TestObjects/src/classes_def.xml`
```diff
@@ -5,6 +5,7 @@
 <class name="edmtest::IntProduct" ClassVersion="10">
  <version ClassVersion="10" checksum="4286676158"/>
 </class>
+<class name="std::pair<edmtest::IntProduct, edm::RandomNumberGeneratorState>"/>
 <class name="edmtest::MaybeUninitializedIntProduct" ClassVersion="3">
  <version ClassVersion="3" checksum="3324182225"/>
 </class>
```

So I expected the bot to request a new signature only from the `core` category.

jfernan2 commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @mandrenguyen, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
